state_tracker.go

```diff
@@ -137,31 +176,93 @@ func (s *StateTracker) UpdateLastResumableBinlogPositionForTargetVerifier(pos my
 	s.lastStoredBinlogPositionForTargetVerifier = pos
 }
 
-func (s *StateTracker) UpdateLastSuccessfulPaginationKey(table string, paginationKey uint64) {
+func (s *StateTracker) RegisterBatch(table string, index uint64, startPaginationKey uint64, endPaginationKey uint64) {
+	s.CopyRWMutex.Lock()
+	defer s.CopyRWMutex.Unlock()
+
+	if _, ok := s.batchProgress[table]; !ok {
+		s.batchProgress[table] = make(map[uint64]*BatchProgress)
+	}
+
+	s.batchProgress[table][index] = &BatchProgress{
+		StartPaginationKey:  startPaginationKey,
+		EndPaginationKey:    endPaginationKey,
+		LatestPaginationKey: startPaginationKey,
+		Completed:           false,
+	}
+}
+
+func (s *StateTracker) UpdateBatchPosition(table string, index uint64, latestPaginationKey uint64) {
+	s.CopyRWMutex.Lock()
+
+	logger := s.logger.WithField("table", table).WithField("batchId", index)
+
+	_, tableFound := s.batchProgress[table]
+	if !tableFound {
+		logger.Error("tried to mark non-existing batch as completed")
+		return
+	}
+
+	batch, batchFound := s.batchProgress[table][index]
+	if !batchFound {
+		logger.Error("tried to mark non-existing batch as completed")
+		return
+	}
+
+	newLatestPaginationKey := uint64(math.Min(float64(latestPaginationKey), float64(batch.EndPaginationKey)))
+	s.updateSpeedLog(newLatestPaginationKey - batch.LatestPaginationKey)
+
+	// Set latest key to EndPaginationKey if we are ahead due to a fragmented PK
+	batch.LatestPaginationKey = newLatestPaginationKey
+
+	if batch.EndPaginationKey == batch.LatestPaginationKey {
+		logger.Debug("marking batch as completed")
+		batch.Completed = true
+		batch.completedPercentage = 100
+		s.CopyRWMutex.Unlock()
```

Comment:
Can we move the `s.CopyRWMutext.Unlock` in both the `if` and `else` case and use `defer` instead to keep it consistent and in case of dangling locks?

Reply:
We can't since we want to release and read in this method. But I can remove one of them and add `defer` to address your dangling point.